File: scripts/data_loading/1_4_2_uis_loader_SDG.R
################################################################# #
# 1.4.2 UIS Data Loader - SDG
################################################################# #
# This script contains the code for downloading and cleaning SDG data
#---------------------------------------------------------------- -
# 0. Importing Modules and Defining Environment-----
#---------------------------------------------------------------- -


#---------------------------------------------------------------- -
# 1. Downloading OPRI Data-----
#---------------------------------------------------------------- -
# Specify the URL of the zip file
url<- "https://apimgmtstzgjpfeq2u763lag.blob.core.windows.net/content/MediaLibrary/bdds/SDG_092023.zip"

# Specify the directory to unzip the files into (make sure this directory exists or create it)
dir.create(fldr_uis)
dir.create(str_c(fldr_uis,'/raw'))

# Specify the local path to save the zip file
destfile <- str_c(fldr_uis,'/raw/OPRI_092023.zip')

# Download the file
download.file(url, destfile, mode="wb")

# Unzip the file
unzip(destfile, exdir=str_c(fldr_uis,'/raw'))

#---------------------------------------------------------------- -
# 2. Data reading-----
#---------------------------------------------------------------- -

# Read SDG NATIONAL (Countries) Data
sdg_n<-read_csv(str_c(fldr_uis,'/raw/SDG_092023/SDG_DATA_NATIONAL.csv')) |> 
  rename(indicator = indicator_id,
         isoalpha3 = country_id) |> 
  # Adding source
  mutate(source = 'UNESCO-UIS')

# Read SDG REGIONAL (Region) Data
sdg_r<-read_csv(str_c(fldr_uis,'/raw/SDG_092023/SDG_DATA_REGIONAL.csv'))|> 
  rename(indicator = indicator_id,
         isoalpha3 = region_id) |> 
  # Adding source
  mutate(source = 'UNESCO-UIS')

# Binding national and regional datasets
sdg<-sdg_n |> 
  bind_rows(sdg_r)

# Reading SDG metadata (Labels)
sdg_dic<-read_csv(str_c(fldr_uis,'/raw/SDG_092023/SDG_LABEL.csv')) |> 
  rename(indicator = INDICATOR_ID,
         label_en = INDICATOR_LABEL_EN) |> 
  mutate(collection = 'International Organizations',
         theme_en = '',
         theme_es = '',
         resource = 'Unesco Institute for Statistics UIS',
         description_es = '',
         description_en = '',
         valuetype = 'level',
         label_es = label_en)
  
# ADD REGIONAL 
# OECD
sdg_oecd<-sdg |> 
  filter(isoalpha3 %in% oecd_list_iso3) |> 
  group_by(indicator,year) |> 
  summarise(value = mean(value,na.rm=T)) |> 
  mutate(isoalpha3='OECD')

# LAC
sdg_lac<-sdg |> 
  filter(isoalpha3 %in% lac_list_iso3) |> 
  group_by(indicator,year) |> 
  summarise(value = mean(value,na.rm=T)) |> 
  mutate(isoalpha3='LAC')

sdg<-
  sdg|> 
  filter(!(isoalpha3 %in% c('LAC','OECD'))) |> 
  #filter(isoalpha3 %in% lac_list_iso3) |> 
  bind_rows(sdg_lac,sdg_oecd)|> 
  mutate(last_saved = today())

# Exporting simple clean
dir.create(str_c(fldr_uis,'/raw_clean'))
write_csv(sdg,
          str_c(fldr_uis,'/raw_clean/indicators_unesco_uis_sdg.csv'))
write_csv(sdg_dic,
          str_c(fldr_uis,'/raw_clean/dictionary_unesco_uis_sdg.csv'))

rm(sdg_n,sdg_r,sdg_lac,sdg_oecd)
#---------------------------------------------------------------- -
# 3. Data cleaning-----
#---------------------------------------------------------------- -
sdg_dic<-read_csv(str_c(fldr_uis,'/raw_clean/dictionary_unesco_uis_sdg.csv'))
sdg<-read_csv(str_c(fldr_uis,'/raw_clean/indicators_unesco_uis_sdg.csv')) |> 
  # Ensuring that the indicators in the dictionary are in the dataset (deleting redundancies)
  filter(indicator %in% sdg_dic$indicator) |>  # no rows removed
  left_join(sdg_dic |> select(indicator,label_en,valuetype)) |> 
  # This next filter is useless as we define everything as level...
  mutate(value = if_else(valuetype=='pct',value/100,value))

# Exporting clean
dir.create(str_c(fldr_uis,'/clean'))
write_csv(sdg,
          str_c(fldr_uis,'/clean/indicators_unesco_uis_sdg.csv'))

File: scripts/data_loading/1_4_2_uis_loader_SDG.py
# DOWNLOADING DATA
# Bin #1
#!/usr/bin/python3
import requests
from bs4 import BeautifulSoup
from pprint import pprint
import zipfile
from urllib.request import urlopen
import io
import pandas as pd
import numpy as np
import dotenv
import os 
# dotenv.load_dotenv('../.env')
# sclbucket = os.getenv('sclbucket')
# scldatalake = f"s3://{sclbucket}"
from datetime import date
todays_date = date.today()
fldr_main='/Users/jmargitic/Library/CloudStorage/OneDrive-SharedLibraries-Inter-AmericanDevelopmentBankGroup/SCL EDU_gasto_inteligente - General/smart_spending_diag_lac_tool'

# Bin #2
url = "https://apimgmtstzgjpfeq2u763lag.blob.core.windows.net/content/MediaLibrary/bdds/SDG_092023.zip"
fldr_date=str(todays_date.year)+'_'+str(todays_date.month)
extract_dir = '/temp/'+str(fldr_date)+'/'

r = requests.get(url)
z = zipfile.ZipFile(io.BytesIO(r.content))
z.extractall(fldr_main+extract_dir)

### LAC countries
countries = pd.read_excel(fldr_main+'/input/IADB_country_codes_admin_0.xlsx', engine='openpyxl')

### read data
data = pd.read_csv(fldr_main+extract_dir+'/SDG_092023/SDG_DATA_NATIONAL.csv')
data = data.rename(columns={'indicator_id':'indicator', 'country_id': 'isoalpha3', 
                            'year':'year', 'value': 'value'})

data['source'] = 'UNESCO-UIS'
data = data[['isoalpha3', 'year', 'indicator', 'value', 'source']]
data = data[data.isoalpha3.isin(countries.isoalpha3)]

### read metadata
dictionary = pd.read_csv(fldr_main+extract_dir+'/SDG_092023/SDG_LABEL.csv')
dictionary = dictionary.rename(columns={'INDICATOR_ID':'indicator', 'INDICATOR_LABEL_EN': 'label_en'})
dictionary['collection'] = 'International Organizations'
dictionary['theme_es'] = ''
dictionary['theme_en'] = ''
dictionary['resource'] = 'Unesco Institute for Statistics UIS'
dictionary['description_es'] = ''
dictionary['description_en'] = ''
dictionary['valuetype'] = 'level'
dictionary['label_es'] = dictionary['label_en']
dictionary = dictionary[['collection', 'resource','theme_es','theme_en','label_es', 'label_en','description_en','description_es','indicator','valuetype']]
dictionary.head()

### exporting
data.to_csv(fldr_main+f'/temp/uis_raw_{fldr_date}/indicators_unesco_uis_sdg.csv', index=False)
dictionary.to_csv(fldr_main+f'/temp/uis_raw_{fldr_date}//dictionary_unesco_uis_sdg.csv', index=False)


# cleaning data
data = pd.read_csv(fldr_main+f'/temp/uis_raw_{fldr_date}/indicators_unesco_uis_sdg.csv')
dictionary = pd.read_csv(fldr_main+f'/temp/uis_raw_{fldr_date}//dictionary_unesco_uis_sdg.csv', sep=',')
data = data[data.indicator.isin(dictionary.indicator)]
data.loc[(data.indicator.isin(dictionary[dictionary.valuetype=='pct'].indicator)), 'value'] = data.loc[(data.indicator.isin(dictionary[dictionary.valuetype=='pct'].indicator)), 'value']/100
# export
data.to_csv(fldr_main+'/temp/indicators_unesco_uis_sdg.csv', index=False)

# Filtering indicators
data = data[data.indicator.isin([
"XUNIT.GDPCAP.1.FSGOV.FFNTR", #
"XUNIT.GDPCAP.2T3.FSGOV.FFNTR", #
"XUNIT.GDPCAP.5T8.FSGOV.FFNTR", #
"XUNIT.PPPCONST.1.FSGOV.FFNTR", #
"XUNIT.PPPCONST.2T3.FSGOV.FFNTR", #
"XUNIT.PPPCONST.5T8.FSGOV.FFNTR", #
"XGDP.FSGOV" #
])]

data.to_csv(fldr_main+'/temp/indicators_unesco_uis_sdg_FILTERED.csv', index=False)


